src/main/java/com/gregtechceu/gtceu/common/machine/electric/PumpMachine.java

```diff
@@ -128,59 +147,274 @@ private int getMaxPumpRange() {
+    /**
+     * Returns a list of directions, starting with Up and then horizontal directions with the directions most matching
+     * the vector first.
+     */
+    private List<Direction> biasedInVecDirections(RandomSource randomSource, Vec3i vec, boolean goUp) {
+        List<Direction> searchList = new ArrayList<>();
+        if (goUp) {
+            searchList.add(Direction.UP);
+        }
+
+        ObjectArrayList<Direction.Axis> axes = new ObjectArrayList<>();
+        int zValue = Math.abs(vec.getZ());
+        int xValue = Math.abs(vec.getX());
+        if (zValue > xValue) {
+            axes.add(Direction.Axis.Z);
+            axes.add(Direction.Axis.X);
+        } else if (zValue < xValue) {
+            axes.add(Direction.Axis.X);
+            axes.add(Direction.Axis.Z);
+        } else {
+            axes.add(Direction.Axis.Z);
+            axes.add(Direction.Axis.X);
+            Util.shuffle(axes, randomSource);
+        }
+
+        Direction lastDirection = null;
+        for (int i = 0; i < 2; i++) {
+            Direction.Axis axis = axes.get(i);
+            int value;
+            if (axis.equals(Direction.Axis.Z)) {
+                value = vec.getZ();
+            } else {
+                value = vec.getX();
+            }
+
+            Direction direction;
+            if (value < 0) {
+                direction = Direction.fromAxisAndDirection(axis, Direction.AxisDirection.NEGATIVE);
+            } else if (value > 0) {
+                direction = Direction.fromAxisAndDirection(axis, Direction.AxisDirection.POSITIVE);
+            } else {
+                direction = Direction.fromAxisAndDirection(axis, Util.getRandom(Direction.AxisDirection.values(), randomSource));
+            }
+            searchList.add(direction);
+            if (i == 0) {
+                lastDirection = direction.getOpposite();
+            } else {
+                searchList.add(direction.getOpposite());
+            }
+
+        }
+        searchList.add(lastDirection);
+
+        return searchList;
     }
 
-    private void updateQueueState(int blocksToCheckAmount) {
-        BlockPos selfPos = getPos().below(pumpHeadY);
+    private record PumpQueue(Queue<Deque<BlockPos>> queue, FluidType fluidType) {}
 
-        for (int i = 0; i < blocksToCheckAmount; i++) {
-            BlockPos checkPos = null;
-            int amountIterated = 0;
-            do {
-                if (checkPos != null) {
-                    blocksToCheck.push(checkPos);
-                    amountIterated++;
+    private record SearchResult(BlockPos pos, boolean isSource) {}
+
+    /**
+     *
+     */
+    @Nullable
+    private SearchResult searchNext(Level level, BlockPos headPosBelow, BlockPos searchHead, FluidType fluidType, int maxPumpRange, boolean goUp, Set<BlockPos> checked) {
+        // Vector from the pump head to the search head, so points in the direction away from the pump head
+        Vec3i subVec = searchHead.subtract(headPosBelow);
```

Comment:
just store them as BlockPos, makes it easier to see what it is

Reply:
Sorry, what do you mean with this exactly? Store what as BlockPos? Or you mean I should get rid of the returning the record? I'm used to languages with multiple return types or cheap structs so that's why I did it this way

Reply:
instead of `subVec` being of type Vec3i it should be of type BlockPos

Reply:
Ah but that's on purpose because it doesn't represent a place in the world, it's just the vector between two block positions that points away from the pump. You want me to still change it?

Reply:
meh, it's fine